bin/yb-ctl

```diff
@@ -726,30 +775,39 @@ class ClusterControl:
 
         This assumes you will have called set_master_addresses already!
         """
-        if len(self.options.placement_info) > 0:
-            yb_admin_binary_path = self.options.get_binary_path("yb-admin")
-            cmd = [yb_admin_binary_path, "--master_addresses", self.options.master_addresses,
-                   "modify_placement_info", self.options.placement_info_raw,
-                   str(self.options.replication_factor)]
-            wait_count = 0
-            while wait_count < MAX_WAIT_ITERS:
-                try:
-                    subprocess.check_call(cmd)
-                    print("Successfully modified placement info.")
-                    return
-                except subprocess.CalledProcessError:
-                    wait_count += 1
-                    time.sleep(SLEEP_TIME_IN_SEC)
+        if not self.options.placement_info:
+            return
+
+        yb_admin_binary_path = self.options.get_binary_path("yb-admin")
+        cmd = [yb_admin_binary_path, "--master_addresses", self.options.master_addresses,
+               "modify_placement_info", self.options.placement_info_raw,
+               str(self.get_replication_factor())]
+
+        def call_modify_placement_info():
+            with open(os.devnull, "wb") as devnull:
+                subprocess.check_call(cmd, stderr=devnull)
+            logging.info("Successfully modified placement info.")
+            return True
+        if not retry_fn_with_timeout(call_modify_placement_info, self.options.timeout):
             raise RuntimeError("Could not modify placement info for the cluster.")
 
     def get_number_of_servers(self, daemon_type):
         # When the cluster is being created, YB data directory may not be ready on disk.
         # In that case, use replication_factor to determine number of servers.
         if self.creating_cluster:
-            return self.options.replication_factor
+            return self.get_replication_factor()
         else:
+            drives = self.get_drive_paths()
+            if not drives:
+                return 0
             return len(glob.glob("{}/*/{}/yb-data/{}".format(
-                self.options.cluster_base_dir, self.options.drives[0], daemon_type)))
+                self.options.cluster_base_dir, drives[0], daemon_type)))
```

Comment:
Are we assuming that the list of node ids represented by these directories is contiguous? Or should we perhaps extract the max node id from the directory names?

Reply:
We are indeed assuming this and it is supposed to be how we allocate them when the user adds new nodes.